### PR TITLE
Workload monitoring and analysis

### PR DESCRIPTION
**Pull request by cedricmendelin**
In this PR, the functionality of monitoring in Polypheny-DB is added. The first use case for monitoring in Polypheny will be Query Monitoring, which will already be prepared but not used.

With the merge of temperature aware partitioning and the new routing approach, the monitoring of Queries will be added in a second step.

The monitoring can be easily used for other use cases and any kind of information can be monitored. So far, the monitored events will be analyzed, and data points will be generated. These data points will be persistent to a MapDB repository.

In the future, also other repositories could be made available, like InfluxDB Repository.

### CODE REVIEW

File: core/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueue.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+     * @param eventData the event data which will be queued.
+     */
+    void queueEvent( MonitoringEvent eventData );
+
+    /**
+     * Display current number of elements in queue
+     *
+     * @return Current numbe of elements in Queue
+     */
+    long getNumberOfElementsInQueue();
```
Comment (cedricmendelin) [inline core/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueue.java:43]: @hennlo should we make an int as return type and only return number of elements in queue. So basically remove info of getNumberOfProcessedEvents?

File: core/src/main/java/org/polypheny/db/monitoring/events/BaseEvent.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    private long recordedTimestamp;
+
+
+    public BaseEvent() {
+        setEventType( eventType );
+        recordedTimestamp = getCurrentTimestamp();
+    }
+
+
+    public void setEventType( String eventType ) {
```
Comment (cedricmendelin) [inline core/src/main/java/org/polypheny/db/monitoring/events/BaseEvent.java:41]: @hennlo You once added this eventType? Should we remove it and just use the implementing class as type? I think I already removed some part of it

File: dbms/src/main/java/org/polypheny/db/transaction/TransactionImpl.java [hunk 3]
```diff
@@ -79,15 +80,13 @@
 
     @Getter
     private final boolean analyze;
-
+    private final AtomicLong statementCounter = new AtomicLong();
```
Comment (cedricmendelin) [inline dbms/src/main/java/org/polypheny/db/transaction/TransactionImpl.java:83]: @hennlo did we add this? Should we remove it?

File: core/src/main/java/org/polypheny/db/monitoring/events/BaseEvent.java (hunk 2 above)
Comment (hennlo) [reply core/src/main/java/org/polypheny/db/monitoring/events/BaseEvent.java:41]: This is still required, the implementing ClassType e.g. could be DML
The EventType is a String translation form SQLKind: INSERT/UPDATE/DELETE and is therefore more specific.

I would propose to keep the functionality

File: core/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueue.java (hunk 1 above)
Comment (hennlo) [reply core/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueue.java:43]: Okay I guess int is suffcient if we change the return from:
`queueIds.size()`
to
`monitoringJobQueue.size()`

but I would still keep `getNumberOfProcessedEvents(boolean all)`
That is quite useful for gathering statistics.

File: dbms/src/main/java/org/polypheny/db/transaction/TransactionImpl.java (hunk 3 above)
Comment (hennlo) [reply dbms/src/main/java/org/polypheny/db/transaction/TransactionImpl.java:83]: @cedricmendelin  no this was added by @vogti

File: core/src/main/java/org/polypheny/db/monitoring/events/BaseEvent.java (hunk 2 above)
Comment (cedricmendelin) [reply core/src/main/java/org/polypheny/db/monitoring/events/BaseEvent.java:41]: But here we are not talking about sql or something like that. This is the base event of our monitoring.

If you want to have something like that for query events we can add it there, what do you think?

To be honest, I do not see any benefit of the string in the base event, we can always just get the implementing class. If a event need kind of type (which could even be different within the same class), i would implement it directly on this event and not on the base event.

Comment (hennlo) [reply core/src/main/java/org/polypheny/db/monitoring/events/BaseEvent.java:41]: Ah okay, yes you're absolutely right, I looked at the wrong part!
Yes in that case eventType is only necessary in StataementEvent or Subsclasses.

I would propose StatementEvent.

File: dbms/src/main/java/org/polypheny/db/transaction/TransactionImpl.java (hunk 3 above)
Comment (vogti) [reply dbms/src/main/java/org/polypheny/db/transaction/TransactionImpl.java]: You're right, that is no longer necessary. I've removed it.

File: dbms/src/main/java/org/polypheny/db/PolyphenyDb.java [hunk 4]
```diff
@@ -261,9 +260,6 @@ public void join( final long millis ) throws InterruptedException {
             throw new RuntimeException( "Something went wrong while initializing index manager.", e );
         }
 
-        // Call DockerManager once to remove old containers
-        DockerManager.getInstance();
-
```
Comment (vogti) [inline dbms/src/main/java/org/polypheny/db/PolyphenyDb.java:266]: Why removing this call?

File: webui/src/main/java/org/polypheny/db/webui/Crud.java [hunk 5]
```diff
@@ -248,6 +248,58 @@
     }
```
Comment (vogti) [inline webui/src/main/java/org/polypheny/db/webui/Crud.java:250]: Why reordering the methods in this class?

File: dbms/src/main/java/org/polypheny/db/processing/AbstractQueryProcessor.java [hunk 6]
```diff
@@ -140,20 +140,59 @@
 @Slf4j
 public abstract class AbstractQueryProcessor implements QueryProcessor {
```
Comment (vogti) [inline dbms/src/main/java/org/polypheny/db/processing/AbstractQueryProcessor.java:141]: Why reordering methods and attributes in this class? Has there been any other changes?

File: information/src/main/java/org/polypheny/db/information/InformationDuration.java [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
```
Comment (vogti) [inline information/src/main/java/org/polypheny/db/information/InformationDuration.java:1]: Why were some methods and attributes rearranged in this class?

File: monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+
+    @Override
+    public long getNumberOfProcessedEvents() {
+        return processedEvents;
+    }
+
+    // endregion
+
+    // region private helper methods
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java:150]: ?

File: monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    /**
+     * Ctor will automatically start the background task for consuming the queue.
+     */
+    public MonitoringQueueImpl( @NonNull MonitoringRepository repository ) {
+        this( true, repository );
+    }
+
+    // endregion
+
+    // region public methods
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java:95]: ?

File: monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+    /**
+     * Processed events since restart.
+     */
+    private long processedEvents;
+
+    // endregion
+
+    // region ctors
+
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java:68]: ?

File: monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+    private Optional<MonitoringEvent> getNextJob() {
+        if ( monitoringJobQueue.peek() != null ) {
+            return Optional.of( monitoringJobQueue.poll() );
+        }
+        return Optional.empty();
+    }
+
+    // endregion
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java:205]: ?

File: monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+import org.polypheny.db.util.background.BackgroundTaskManager;
+
+/**
+ * MonitoringQueue implementation which stores the monitoring jobs in a
+ * concurrentQueue and will process them with a background worker task.
+ */
+@Slf4j
+public class MonitoringQueueImpl implements MonitoringQueue {
+
+    // region private fields
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java:47]: Why did you add these code arrangement comments here?

File: monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+    // region public methods
+
+
+    @Override
+    protected void finalize() throws Throwable {
+        super.finalize();
+        if ( backgroundTaskId != null ) {
+            BackgroundTaskManager.INSTANCE.removeBackgroundTask( backgroundTaskId );
+        }
+    }
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java:104]: Finalizers are deprecated and should not be used.
https://wiki.sei.cmu.edu/confluence/display/java/MET12-J.+Do+not+use+finalizers

File: monitoring/src/test/java/org/polypheny/db/monitoring/core/MonitoringQueueImplIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        // assert
+
+        val result = sut.getAllDataPoints( QueryDataPoint.class );
+        Assertions.assertEquals( 15, result.size() );
+
+        // cleanup
+        try {
+            queueWriteService.finalize();
```
Comment (vogti) [inline monitoring/src/test/java/org/polypheny/db/monitoring/core/MonitoringQueueImplIntegrationTest.java:74]: Calling Object.finalize() explicitly is a very bad idea, as it can result in objects being placed in an inconsistent state. Even in a test case.

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DMLEventAnalyzer.java [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+            getDurationInfo( metric, "Index Lookup Rewrite", duration );
+            getDurationInfo( metric, "Constraint Enforcement", duration );
+            getDurationInfo( metric, "Implementation Caching", duration );
+            getDurationInfo( metric, "Index Update", duration );
+            getDurationInfo( metric, "Routing", duration );
+            getDurationInfo( metric, "Planning & Optimization", duration );
+            getDurationInfo( metric, "Implementation", duration );
+            getDurationInfo( metric, "Locking", duration );
+        } catch ( Exception e ) {
+            log.debug( "could not deserialize of get duration info" );
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DMLEventAnalyzer.java:72]: This more or less ignores the exception...

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DMLEventAnalyzer.java [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        }
+    }
+
+
+    private static void getDurationInfo( DMLDataPoint dmlMetric, String durationName, InformationDuration duration ) {
+        try {
+            long time = duration.getDuration( durationName );
+            dmlMetric.getDataElements().put( durationName, time );
+        } catch ( Exception e ) {
+            log.debug( "could no find duration:" + durationName );
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DMLEventAnalyzer.java:82]: This more or less ignores the exception...

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/QueryEventAnalyzer.java [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+            getDurationInfo( metric, "Index Lookup Rewrite", duration );
+            getDurationInfo( metric, "Constraint Enforcement", duration );
+            getDurationInfo( metric, "Implementation Caching", duration );
+            getDurationInfo( metric, "Index Update", duration );
+            getDurationInfo( metric, "Routing", duration );
+            getDurationInfo( metric, "Planning & Optimization", duration );
+            getDurationInfo( metric, "Implementation", duration );
+            getDurationInfo( metric, "Locking", duration );
+        } catch ( Exception e ) {
+            log.debug( "could not deserialize of get duration info" );
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/QueryEventAnalyzer.java:70]: This more or less ignores the exception...

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/QueryEventAnalyzer.java [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }
+
+
+    private static void getDurationInfo( QueryDataPoint queryMetric, String durationName, InformationDuration duration ) {
+        try {
+            long time = duration.getDuration( durationName );
+            queryMetric.getDataElements().put( durationName, time );
+        } catch ( Exception e ) {
+            log.debug( "could no find duration:" + durationName );
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/QueryEventAnalyzer.java:80]: This more or less ignores the exception...

File: monitoring/src/test/java/org/polypheny/db/monitoring/core/MonitoringQueueImplIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    private List<QueryEvent> createQueryEvent( int number ) {
+        val result = new ArrayList<QueryEvent>();
+
+        for ( int i = 0; i < number; i++ ) {
+            val event = new QueryEvent();
+            event.setRouted( null );
+            event.setSignature( null );
+            event.setStatement( Mockito.mock( Statement.class ) );
+            event.setDescription( UUID.randomUUID().toString() );
+            event.setExecutionTime( (long) Math.random() * 1000L );
```
Comment (vogti) [inline monitoring/src/test/java/org/polypheny/db/monitoring/core/MonitoringQueueImplIntegrationTest.java:92]: This is always 0....

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DMLEventAnalyzer.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.polypheny.db.information.InformationDuration;
+import org.polypheny.db.monitoring.events.DMLEvent;
+import org.polypheny.db.monitoring.events.metrics.DMLDataPoint;
+import org.polypheny.db.rel.RelNode;
+import org.polypheny.db.rel.RelRoot;
+
+@Slf4j
+public class DMLEventAnalyzer {
+    // TODO: Bis jetzt sind die Klassen mehr oder weniger identisch. Ist das einfach vorbereitet für später oder wie?
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DMLEventAnalyzer.java:30]: Please comment in English.

File: monitoring/src/main/java/org/polypheny/db/monitoring/ui/MonitoringServiceUiImpl.java [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+
+@Slf4j
+public class MonitoringServiceUiImpl implements MonitoringServiceUi {
+
+    private final MonitoringRepository repo;
+    private final MonitoringQueue queue;
+    private InformationPage informationPage;
+
+
+    // TODO: die Abhänigkeit zur Queue ist nicht wirklich optimal, aber lassen wir vielleicht mal so stehen
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/ui/MonitoringServiceUiImpl.java:49]: Please comment in English.

File: core/src/main/java/org/polypheny/db/config/RuntimeConfig.java [hunk 22]
```diff
@@ -530,75 +553,80 @@ public boolean getBoolean() {
...
+        configManager.getConfig( key ).setBoolean( value );
+    }
+
+
     public BigDecimal getDecimal() {
         return configManager.getConfig( key ).getDecimal();
     }
 
 
-    public double getDouble() {
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/config/RuntimeConfig.java:538]: Why have you rearranged the methods in this class?

File: monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java (hunk 12 above)
Comment (cedricmendelin) [reply monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java:47]: Thats a feature of IntelliJ to nicely show structures of classes. We can remove them again if you find it disturbing:
image

Comment (vogti) [reply monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java]: Of course, if you find them useful, we can put them in the code. For such small classes I don't really see the benefit, but for me it's fine to have the comments in the code if you find them useful.

File: dbms/src/main/java/org/polypheny/db/PolyphenyDb.java (hunk 4 above)
Comment (hennlo) [reply dbms/src/main/java/org/polypheny/db/PolyphenyDb.java]: @cedricmendelin  maybe this  did happen during merge?
I am not aware that I removed it.

But I fixed it now.

File: information/src/main/java/org/polypheny/db/information/InformationDuration.java [hunk 23]
```diff
@@ -62,6 +61,15 @@ public Duration get( final String name ) {
     }
 
 
+    public long getNanoDuration( final String name ) {
+        Duration child = this.get( name );
+        if ( child != null ) {
+            return child.duration;
+        }
+        return 0;
```
Comment (vogti) [inline information/src/main/java/org/polypheny/db/information/InformationDuration.java:69]: It would be better to throw an Exception. Such "default" values are always dangerous....

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DmlEventAnalyzer.java [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+            getDurationInfo( metric, "Index Lookup Rewrite", duration );
+            getDurationInfo( metric, "Constraint Enforcement", duration );
+            getDurationInfo( metric, "Implementation Caching", duration );
+            getDurationInfo( metric, "Index Update", duration );
+            getDurationInfo( metric, "Routing", duration );
+            getDurationInfo( metric, "Planning & Optimization", duration );
+            getDurationInfo( metric, "Implementation", duration );
+            getDurationInfo( metric, "Locking", duration );
+        } catch ( Exception e ) {
+            log.debug( "could not deserialize of get duration info" );
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DmlEventAnalyzer.java:70]: This hides an error...

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DmlEventAnalyzer.java [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+
+    private static void getDurationInfo( DmlDataPoint dmlMetric, String durationName, InformationDuration duration ) {
+        try {
+            long time = duration.getNanoDuration( durationName );
+            dmlMetric.getDataElements().put( durationName, time );
+        } catch ( Exception e ) {
+
+            log.debug( "could no find duration:{}", durationName );
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/DmlEventAnalyzer.java:81]: This hides an error...

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/QueryEventAnalyzer.java [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+        }
+    }
+
+
+    private static void getDurationInfo( QueryDataPoint queryMetric, String durationName, InformationDuration duration ) {
+        try {
+            long time = duration.getNanoDuration( durationName );
+            queryMetric.getDataElements().put( durationName, time );
+        } catch ( Exception e ) {
+            log.debug( "could no find duration: {}", durationName );
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/analyzer/QueryEventAnalyzer.java:81]: This hides an error...

File: core/src/main/java/org/polypheny/db/monitoring/events/MonitoringEvent.java [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+public interface MonitoringEvent {
+
+    UUID getId();
+
+    Timestamp getRecordedTimestamp();
+
+    /**
+     * @param <T> Defined Class Types which will be generated from the event.
+     * The analyze method will create the list of metrics.
+     * @return
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/events/MonitoringEvent.java:38]: Please fix JavaDoc

File: core/src/main/java/org/polypheny/db/monitoring/events/MonitoringEvent.java [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+     */
+    <T extends MonitoringDataPoint> List<Class<T>> getMetrics();
+
+    /**
+     * @param <T> defined Class Types which will optionally be generated from the event.
+     * The analyze method will attach the optional metrics.
+     * @return
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/events/MonitoringEvent.java:45]: Please fix JavaDoc

File: core/src/main/java/org/polypheny/db/monitoring/persistence/MonitoringRepository.java [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * Initialized the repository, might need some configuration beforehand.
+     */
+    void initialize();
+
+    /**
+     * Persist given monitoring metric.
+     *
+     * @param dataPoint
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/persistence/MonitoringRepository.java:37]: Please add description

File: core/src/main/java/org/polypheny/db/monitoring/persistence/MonitoringRepository.java [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    void persistDataPoint( MonitoringDataPoint dataPoint );
+
+    /**
+     * Get all data for given monitoring persistent type.
+     *
+     * @param dataPointClass
+     * @param <T>
+     * @return
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/persistence/MonitoringRepository.java:46]: Please fix JavaDoc

File: core/src/main/java/org/polypheny/db/monitoring/persistence/MonitoringRepository.java [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    <T extends MonitoringDataPoint> List<T> getAllDataPoints( Class<T> dataPointClass );
+
+    /**
+     * Get data before specified timestamp for given monitoring persistent type.
+     *
+     * @param dataPointClass
+     * @param <T>
+     * @return
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/persistence/MonitoringRepository.java:55]: Please fix JavaDoc

File: core/src/main/java/org/polypheny/db/monitoring/persistence/MonitoringRepository.java [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    <T extends MonitoringDataPoint> List<T> getDataPointsBefore( Class<T> dataPointClass, Timestamp timestamp );
+
+    /**
+     * Get data after specified timestamp for given monitoring persistent type.
+     *
+     * @param dataPointClass
+     * @param <T>
+     * @return
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/persistence/MonitoringRepository.java:64]: Please fix JavaDoc

File: core/src/main/java/org/polypheny/db/monitoring/ui/MonitoringServiceUi.java [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+public interface MonitoringServiceUi {
+
+    void initializeInformationPage();
+
+    /**
+     * Will add new section to monitoring information page for the specified
+     * MonitoringPersistentData type and register the refresh function to read from repository.
+     *
+     * @param metricClass
+     * @param <T>
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/ui/MonitoringServiceUi.java:34]: Please fix JavaDoc

File: core/src/main/java/org/polypheny/db/monitoring/core/MonitoringService.java [hunk 34]
```diff
@@ -0,0 +1,66 @@
...
+     * @param <T> The type parameter for the event, which will implement MonitoringEventData
+     */
+    <T extends MonitoringEvent> void monitorEvent( T eventData );
+
+    /**
+     * Get all data for given monitoring persistent type.
+     *
+     * @param dataPointClass
+     * @param <T>
+     * @return
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/core/MonitoringService.java:44]: Please fix JavaDoc

File: core/src/main/java/org/polypheny/db/monitoring/core/MonitoringService.java [hunk 35]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    <T extends MonitoringDataPoint> List<T> getAllDataPoints( Class<T> dataPointClass );
+
+    /**
+     * Get data before specified timestamp for given monitoring persistent type.
+     *
+     * @param dataPointClass
+     * @param <T>
+     * @return
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/core/MonitoringService.java:53]: Please fix JavaDoc

File: core/src/main/java/org/polypheny/db/monitoring/core/MonitoringService.java [hunk 36]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    <T extends MonitoringDataPoint> List<T> getDataPointsBefore( Class<T> dataPointClass, Timestamp timestamp );
+
+    /**
+     * Get data after specified timestamp for given monitoring persistent type.
+     *
+     * @param dataPointClass
+     * @param <T>
+     * @return
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/core/MonitoringService.java:62]: Please fix JavaDoc

File: monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java [hunk 37]
```diff
@@ -0,0 +1,205 @@
...
+    // region ctors
+
+
+    /**
+     * Ctor which automatically will start the background task based on the given boolean
+     *
+     * @param startBackGroundTask Indicates whether the background task for consuming the queue will be started.
+     */
+    public MonitoringQueueImpl( boolean startBackGroundTask, @NonNull MonitoringRepository repository ) {
+        log.info( "write queue service" );
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java:73]: This log message should either be removed or changed to debug

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/DmlEvent.java [hunk 38]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+
+    @Override
+    public List<MonitoringDataPoint> analyze() {
+        try {
+            return Arrays.asList( DmlEventAnalyzer.analyze( this ) );
+        } catch ( Exception e ) {
+            log.error( "Could not analyze dml event" );
+            return Collections.emptyList();
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/DmlEvent.java:47]: This somehow hides an error which potentially causes other errors.

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/QueryEvent.java [hunk 39]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+
+    @Override
+    public List<MonitoringDataPoint> analyze() {
+        try {
+            return Arrays.asList( QueryEventAnalyzer.analyze( this ) );
+        } catch ( Exception e ) {
+            log.error( "Could not analyze query event:" );
+            return Collections.emptyList();
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/QueryEvent.java:47]: This somehow hides an error which potentially causes other errors.

File: monitoring/src/main/java/org/polypheny/db/monitoring/persistence/MapDbRepository.java [hunk 40]
```diff
@@ -0,0 +1,152 @@
...
+                    .collect( Collectors.toList() );
+        }
+
+        return Collections.emptyList();
+    }
+
+
+    @Override
+    public <T extends MonitoringDataPoint> List<T> getDataPointsBefore( @NonNull Class<T> dataPointClass, @NonNull Timestamp timestamp ) {
+        // TODO: not tested yet
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/persistence/MapDbRepository.java:91]: todo?

File: monitoring/src/main/java/org/polypheny/db/monitoring/persistence/MapDbRepository.java [hunk 41]
```diff
@@ -0,0 +1,152 @@
...
+                    .collect( Collectors.toList() );
+        }
+
+        return Collections.emptyList();
+    }
+
+
+    @Override
+    public <T extends MonitoringDataPoint> List<T> getDataPointsAfter( @NonNull Class<T> dataPointClass, @NonNull Timestamp timestamp ) {
+        // TODO: not tested yet
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/persistence/MapDbRepository.java:108]: todo?

File: monitoring/src/main/java/org/polypheny/db/monitoring/ui/MonitoringServiceUiImpl.java [hunk 42]
```diff
@@ -0,0 +1,185 @@
...
+                // Therefore, nothing need to be done for serialVersionID
+                // and neither do we need to hacky set the setAccessible flag for the fields
+                if ( field.getName().equals( "serialVersionUID" ) ) {
+                    continue;
+                }
+
+                try {
+                    field.setAccessible( true );
+                    val value = field.get( element );
+                    row.add( value.toString() );
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/ui/MonitoringServiceUiImpl.java:122]: Why do you need reflections here?

Comment (hennlo) [reply monitoring/src/main/java/org/polypheny/db/monitoring/ui/MonitoringServiceUiImpl.java:122]: @cedricmendelin Can you comment on this

Comment (cedricmendelin) [reply monitoring/src/main/java/org/polypheny/db/monitoring/ui/MonitoringServiceUiImpl.java:122]: Thats a function which will work for every MonitoringDataPoint and will read all fields from the object and shows values on the ui. As this can be arbitrarly objects and the values are not in the Interface, reflection is the way to go.

But all in all, thats just debug output and maybe thats not needed anymore. We could remove it with the corresponding functions in the service, what do you think? Is it a benefit, that you can "register" monitoring data points at the UI or will one do it customized for the data points if needed?

**Review by vogti — APPROVED** [review]
Thx @cedricmendelin and @hennlo for this PR!
